### PR TITLE
Flask backend

### PR DESCRIPTION
**Pull request by ceprio**
Hi almarklein,

Here is a flask backend for flexx. It is implemented with minimum change to the existing files (probably less than 4 lines changes). The changes do not affect current form and function.

The pull request implements two use case: 
1) Stand alone application (see example in howtos/flask_backend.py)
2) In server flexx application (see example in howtos/flask_server.py)

I also added some goodies: A markdown widget and some ui_usage example (I really like to be able the rapidly test simple widgets). The examples are easier to use than the cut and paste from the documentation.

With flask being more popular than tornado, it should help push this project.

### CODE REVIEW

**Comment by ceprio** [comment]
Forgot to mention a few suggestions for improvements and maintability:
1) Code from MainHandler, AppHandler and WSHandler should be moved in _serverHandlers.py. Only RequestHandler and MyWebSocketHandler base class are needed for the backend.
2) The specification of the backend should be passed at run or start. Not to launch or before. This is to separate server functions from client functions.
This should help for the next backend implementation.

File: flexx/app/_flaskhelpers.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+def start_thread():
+    import threading
+    import asyncio
+
+    def flexx_thread():
+        """
+        Function to start a thread containing the main loop of flexx.
+        This is needed as flexx is an asyncio application which is not 
+        compatible with flexx/gevent.
```
Comment (almarklein) [inline flexx/app/_flaskhelpers.py:90]: I think you mean flask/gevent here?

File: flexx/ui/widgets/_markdown.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+"""
+
+Simple example:
+
+.. UIExample:: 200
+
+    # Define data. This can also be generated with the plotly Python library
+    data = [{'type': 'bar',
```
Comment (almarklein) [inline flexx/ui/widgets/_markdown.py:8]: This example looks to be from another file :)

File: flexx/ui/widgets/_markdown.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from ... import app, event
+from . import Widget
+
+app.assets.associate_asset(__name__, 'https://cdnjs.cloudflare.com/ajax/libs/showdown/1.9.1/showdown.min.js')
+
+
+class Markdown(Widget):
+    """ A widget that shows a rendered Markdown content.
+    """
+    CSS = """
```
Comment (almarklein) [inline flexx/ui/widgets/_markdown.py:28]: Really nice addition!

File: flexxamples/howtos/flask_server.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+from flask import Flask, current_app, url_for
+from flask_sockets import Sockets
```
Comment (almarklein) [inline flexxamples/howtos/flask_server.py:2]: Can you add a docstring explaining what this example does? I think the idea is to embed a Flexx app in a Flask application, but some more details about how this works would be nice.

File: flexxamples/ui_usage/combo_box.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+"""
+Simple use of a dropdown,
+"""
```
Comment (almarklein) [inline flexxamples/ui_usage/combo_box.py:3]: Unfinished docstring? Also some commented code further down. Probably better to either revive or remove.

File: flexxamples/howtos/python_side_widget2.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+from flexx import flx
+
```
Comment (almarklein) [inline flexxamples/howtos/python_side_widget2.py:2]: A short docstring here would be nice as well.

File: flexxamples/ui_usage/label.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+"""
+Simple use of a dropdown,
+"""
```
Comment (almarklein) [inline flexxamples/ui_usage/label.py:3]: not a dropdown ;)

**Comment by almarklein** [comment]
Thanks for this! This is looking really nice already.  About these additional improvements. Good points, but let's save these for another PR.

**Comment by ceprio** [comment]
Docstrings... 
I see things like `# doc-export: Example` and `# flake8: noqa` and in string references like this:

    """ Button classes
    
    Simple example:
    
    .. UIExample:: 50
    
        b = ui.Button(text="Push me")
    
    Also see examples: :ref:`buttons.py`.
      """

Do you have some guidelines or references to the tools you are using for documentation? I can then cleanup the docstrings of the files provided. (If I clean up these files, do you intent to add them? I personally find them useful.)

I can do another PR for Markdown and the flask backend only, ui_usage removed and submitted in another PR?

Thanks,

File: flexx/app/_flaskhelpers.py (hunk 1 above)
Comment (ceprio) [reply flexx/app/_flaskhelpers.py:90]: Corrected.

File: flexx/ui/widgets/_markdown.py (hunk 2 above)
Comment (ceprio) [reply flexx/ui/widgets/_markdown.py:8]: Added relevent comments

File: flexx/ui/widgets/_markdown.py (hunk 3 above)
Comment (ceprio) [reply flexx/ui/widgets/_markdown.py:28]: It is usefull.

File: flexxamples/howtos/flask_server.py (hunk 4 above)
Comment (ceprio) [reply flexxamples/howtos/flask_server.py:2]: Done.

File: flexxamples/ui_usage/combo_box.py (hunk 5 above)
Comment (ceprio) [reply flexxamples/ui_usage/combo_box.py:3]: I removed this one. Will add back later if I play more with it. It seem to have issues.

File: flexxamples/howtos/python_side_widget2.py (hunk 6 above)
Comment (ceprio) [reply flexxamples/howtos/python_side_widget2.py:2]: Done.

File: flexxamples/ui_usage/label.py (hunk 7 above)
Comment (ceprio) [reply flexxamples/ui_usage/label.py:3]: Oups, corrected.

**Comment by ceprio** [comment]
I committed my changes to my ceprio/flexx depot. I do not think I can do separate PRs, everything that appears here is from ceprio/master.
Anyway, it should be better now.

**Comment by ceprio** [comment]
Now tested on Python 3.6 and 3.8. 
Note on the side: The asyncio's loop from 3.8 is very unfriendly to threads... Wonder why they did not keep it like 3.6. Anyway it now works.

**Comment by almarklein** [comment]
The `# doc-export: Example` tells the doc builder to embed the app in the docs. The name `Example` is name of the root `flx.Widget` to use.

The `# flake8: noqa` is to tell the linter to skip the line on which this comment is shown.

>  do you intent to add them

Yes, I think that they are useful.

**Comment by ceprio** [comment]
I modified the `# doc-export` in `_markdown.py` to fit the doc builder format. No need for the other files.

**Comment by almarklein** [comment]
Thanks for this very nice contribution!

**Pull request merged**